cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -28,18 +34,137 @@
 public class BasicWorkflowToken implements WorkflowToken {
   private Map<String, Map<String, Long>> mapReduceCounters;
 
+  private Map<String, List<NodeValueEntry>> tokenValueMap = Maps.newHashMap();
+
+  private String nodeName;
+
+  void setCurrentNode(String nodeName) {
+    this.nodeName = nodeName;
+  }
+
+  /**
+   * Merge the other WorkflowToken passed to the method as a parameter
+   * with the WorkflowToken on which the method is invoked.
+   * @param other the other WorkflowToken to be merged
+   */
+  void mergeToken(BasicWorkflowToken other) {
+    Map<String, List<NodeValueEntry>> otherTokenValueMap = other.tokenValueMap;
+
+    for (Map.Entry<String, List<NodeValueEntry>> entry : otherTokenValueMap.entrySet()) {
+      if (!tokenValueMap.containsKey(entry.getKey())) {
+        // Key is newly added to the other WorkflowToken
+        tokenValueMap.put(entry.getKey(), Lists.<NodeValueEntry>newArrayList());
```

Comment:
correct, although it might be better to use ``entry.getValue()`` here followed by ``continue``? This logic is slightly complex. Trying to see how we can remove some complexity.